### PR TITLE
Update Sonobi adapter to accommodate single-size arrays

### PR DESCRIPTION
**Pull request by Studnicky**

### CODE REVIEW

**Comment by mkendall07** [comment]
@Studnicky 
Are the Sonobi test placements still running? Not seeing an ad but we are getting bids back from Sonobi...

**Comment by headerbidding** [comment]
Sonobi adapter works fine here. But we see a 30% discrepancy in revenue - compared to DFP reports.

**Comment by Studnicky** [comment]
@mkendall07 which placements are you using for testing? The ones from the previous pull request? Someone may have archived them.

**Comment by Studnicky** [comment]
@headerbidding - I haven't seen a discrepancy report or test page for your implementation. Please submit discrepancy reports through your account manager, and stay in contact through appropriate channels.

**Comment by mkendall07** [comment]
@Studnicky 
Yes the previous test placements. Are you able to turn them back on?

**Comment by Studnicky** [comment]
Had to make a new order, this one should be permanent and added to our docs for future reference by publishers so they can see it working properly.

**New account style (ad_unit)::**
```
var adUnits = [{
  code: 'div-gpt-ad-1462372786781-2',
  sizes: [[970, 90], [728, 90]],
  bids: [{
    bidder: 'sonobi',
    params: {
      ad_unit: '/7780971/sparks_prebid_LB',
      dom_id: 'div-gpt-ad-1462372786781-2'
    }
  }]
},{
  code: 'div-gpt-ad-1462372786781-1',
  sizes: [300, 250],
  bids: [{
    bidder: 'sonobi',
    params: {
      ad_unit: '/7780971/sparks_prebid_MR',
      dom_id: 'div-gpt-ad-1462372786781-1'
    }
  }]
},{
  code: 'div-gpt-ad-1462372786781-0',
  sizes: [[120, 600], [300, 600], [160, 600]],
  bids: [{
    bidder: 'sonobi',
    params: {
      ad_unit: '/7780971/sparks_prebid_SS',
      dom_id: 'div-gpt-ad-1462372786781-0'
    }
  }]
}];
```

**Old account style (placement_id)::**
```
 var adUnits = [{
  code: 'div-gpt-ad-1462372786781-2',
  sizes: [[970, 90], [728, 90]],
  bids: [{
    bidder: 'sonobi',
    params: {
      placement_id: '1a37836e69eda13e45c6',
      dom_id: 'div-gpt-ad-1462372786781-2'
    }
  }]
},{
  code: 'div-gpt-ad-1462372786781-1',
  sizes: [300, 250],
  bids: [{
    bidder: 'sonobi',
    params: {
      placement_id: 'd869d310bf11c104391e',
      dom_id: 'div-gpt-ad-1462372786781-1'
    }
  }]
},{
  code: 'div-gpt-ad-1462372786781-0',
  sizes: [[120, 600], [300, 600], [160, 600]],
  bids: [{
    bidder: 'sonobi',
    params: {
      placement_id: 'f7a9106052aabd1bb45b',
      dom_id: 'div-gpt-ad-1462372786781-0'
    }
  }]
}];
```

I have not set DFP to bid actively on this account, so you may need to append -test to the placement_id to see bids, as before.

**Comment by Studnicky** [comment]
I apologize in advance, please ignore this for a moment. Another change is pending.

**Comment by Studnicky** [comment]
@mkendall07 :: Updated pull request with error output to help publishers debug their implementations.

File: src/adapters/sonobi.js [hunk 1]
```diff
@@ -13,36 +13,51 @@ var SonobiAdapter = function SonobiAdapter(){
...
+        case (!!o.params.placement_id && o.params.placement_id.length === 0):
+          utils.logError('Sonobi unable to bid: Empty placement_id for ' + o.placementCode);
+          break;
+        case (!!o.params.placement_id):     //  Morpeus style 
+          keyring[o.params.dom_id] = o.params.placement_id + (test ? '-test' : '')  + '|' + sizes;
+          cb_map[o.params.dom_id] = o.placementCode;
+          break;
+        case (!!o.params.ad_unit && o.params.ad_unit.charAt(0) !== '/'):
+          //  DFP docs do not necessarily require leading slash? - add it in if it's not there.
+          o.params.ad_unit = '/' + o.params.ad_unit;
```
Comment (matthewlane) [inline src/adapters/sonobi.js]: Just looking through this and noticed no break statement for this case. Meant to be?

Comment (Studnicky) [inline src/adapters/sonobi.js]: Yes, this is intentional. The action should continue with the fall through.

The jslint library does not recognize this, which is incorrect - this is both valid and intentional.

Comment (matthewlane) [inline src/adapters/sonobi.js]: Oops, I missed the earlier comment about this. Thanks for explaining

**Comment by mkendall07** [comment]
Hey @Studnicky 
Can you explain how the `adunit.code` relates to the `dom_id` ? If I'm a publisher, the `adunit.code` will be specific to my DFP account. If it's required to match an Sonobi identifier it's not going to work. I think I might be missing something. Also is `placement_id` deprecated now?

**Comment by Studnicky** [comment]
@mkendall07 
Many publishers re-use the same adunit.code for multiple slots (i.e. an infinite scroll article scenario).  In these scenarios, we create unique key-vals for our bidder from the ad_unit and dom_id. Our bidder matches them on our side.

In other cases, like static templates where making dynamic pairings doesn't make sense, we use the straightforward placement_id system that matches directly.

The debug conditions were added to help publishers who are struggling with the documentation to figure out what's going wrong with their config without requiring as much attention from our dev team. 

We noticed a while ago that not all publishers use a valid DFP slot code as 'adunit.code', so we require it as a separate parameter. I thought about testing it out and adding it to debug output, but it seemed more prudent to simply request the proper data separately than ask publishers to change config that may affect their other bidders.

Here's a quick test page showing ad_unit config: https://gist.github.com/Studnicky/d6f00d293173fe875ff46ca165ef6e8a

And another using placement_id config:
https://gist.github.com/Studnicky/cfe80a0695621a0ca62996a82c6c6490

**Comment by matthewlane** [comment]
@Studnicky 
Seeing bids come back using the placement_id config you provided, thanks. You mentioned bidding isn't enabled for ad_unit, but is there any way to test for bids using that style of config? Once we confirm that we can merge

**Comment by Studnicky** [comment]
Open the test page, enter this in your js console:

`document.cookie="sbi_test=1; expires=Mon, 30 May 2016 18:14:11 GMT; path=/"`

**Comment by matthewlane** [comment]
@Studnicky Still not getting bids for ad_unit config after entering that line in my console. Anything else I can try to see test data?

**Comment by Studnicky** [comment]
Ooh, cookies on localhost. I'll have to figure something out for you.

**Comment by Studnicky** [comment]
@matthewlane Turned on bidding for the test account.  Should be working now. Let me know if you see nothing after a few refreshes?

**Comment by matthewlane** [comment]
Hmm, still nothing after a few refreshes

**Comment by matthewlane** [comment]
Getting them now, thanks @Studnicky!

**Comment by Studnicky** [comment]
Cool. Before I go tell the rest of the team, does the download page refer to current head, or latest release? Several of our publishers don't build from source and will be dependent on that tool.

**Comment by mkendall07** [comment]
@Studnicky 
The download page is at 0.8.1 currently because of some issues that our internal refactor caused. We don't plan on upgrading it until we release the next tag (0.9.3) to address those issues and also have a buffer of a few days to ensure that there are no lingering issues in master.

**Comment by Studnicky** [comment]
Gotcha, thanks @mkendall07

**Pull request merged**